### PR TITLE
Lessons/blumenkrantz@trailhead.com/apex basics 101

### PR DESCRIPTION
**Pull request by anna-blumenkrantz**
anna/apex basics 101

### CODE REVIEW

File: FAQ.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# FAQ
```
Comment (nchursin) [inline FAQ.md]: Please, remove this file

File: LICENSE [hunk 2]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (nchursin) [inline LICENSE]: Please, remove this file

File: README.md [hunk 3]
```diff
@@ -1,3 +1,6 @@
+<<<<<<< HEAD
```
Comment (nchursin) [inline README.md]: Please, remove these changes

File: force-app/main/default/classes/tasks/BasicTasks.cls [hunk 4]
```diff
@@ -4,8 +4,9 @@ public class BasicTasks {
     // ignore public static.. Focus on variables
     //write your code below this line
     Integer sum = 0;
-
-    return sum;
+    Long aLong = a;
```
Comment (nchursin) [inline force-app/main/default/classes/tasks/BasicTasks.cls:7]: 👍  (done)

File: force-app/main/default/classes/tasks/BasicTasks.cls [hunk 5]
```diff
@@ -14,7 +15,20 @@ public class BasicTasks {
   public static Double basicMathCalculate(Double a, Double b, String operator) {
     //write your code below this line
     Double result = 0;
-
+    switch on operator {
```
Comment (nchursin) [inline force-app/main/default/classes/tasks/BasicTasks.cls:18]: 👍

File: force-app/main/default/classes/tasks/BasicTasks.cls [hunk 6]
```diff
@@ -24,7 +38,11 @@ public class BasicTasks {
   public static Decimal mathPower(Integer a, Integer n) {
     //write your code below this line
     Decimal result = 0;
-
+    Integer j = 0;
+    while (j <= n) {
```
Comment (nchursin) [inline force-app/main/default/classes/tasks/BasicTasks.cls:42]: 👍

File: force-app/main/default/classes/tasks/CollectionsTasks.cls [hunk 7]
```diff
@@ -6,7 +6,13 @@ public class CollectionsTasks {
   public static Integer countZerosInBitList(
     List<Integer> listWithZerosAndUnits
   ) {
-    return 0;
+    integer countZeros = 0;
+    for (Integer element : listWithZerosAndUnits) {
+      if (element == 0) {
```
Comment (nchursin) [inline force-app/main/default/classes/tasks/CollectionsTasks.cls]: Is it possible to do this without this `if`?

File: force-app/main/default/classes/tasks/CollectionsTasks.cls [hunk 8]
```diff
@@ -18,15 +24,31 @@ public class CollectionsTasks {
     Integer sum,
     List<Integer> listWithElementsToCheck
   ) {
-    return false;
+    boolean result = false;
```
Comment (nchursin) [inline force-app/main/default/classes/tasks/CollectionsTasks.cls:43]: 👍

File: force-app/main/default/classes/tasks/CollectionsTasks.cls [hunk 9]
```diff
@@ -18,15 +24,31 @@ public class CollectionsTasks {
...
+    return result;
   }
 
   // use single loop and an arbitrary collection to address same requirements as in sumCanBeReached
   public static Boolean sumCanBeReachedOptimal(
     Integer sum,
     List<Integer> listWithElementsToCheck
   ) {
-    return false;
+    boolean result = false;
```
Comment (nchursin) [inline force-app/main/default/classes/tasks/CollectionsTasks.cls:59]: The idea is perfect! But there's a flaw in the algo. Try running this in apex anonymous:
 
```
Boolean result = CollectionsTasks.sumCanBeReachedOptimal(0, new List<Integer>{0, 1});
System.debug('Should be false. Actual: ' + result);
```

It should print out `Should be false. Actual: false`, but it returns true for your current implementation

File: force-app/main/default/classes/tasks/CollectionsTasks.cls [hunk 10]
```diff
@@ -52,6 +74,13 @@ public class CollectionsTasks {
     List<String> codeTicks,
     List<Integer> priceTicks
   ) {
-    return 0;
+    Integer stockPrice;
```
Comment (nchursin) [inline force-app/main/default/classes/tasks/CollectionsTasks.cls:95]: 👍

**Review by nchursin — CHANGES_REQUESTED** [review]

File: force-app/main/default/classes/tasks/CollectionsTasks.cls (hunk 7 above)
Comment (anna-blumenkrantz) [reply force-app/main/default/classes/tasks/CollectionsTasks.cls]: added no IFs solution, see last commit

File: force-app/main/default/classes/tasks/CollectionsTasks.cls (hunk 9 above)
Comment (anna-blumenkrantz) [reply force-app/main/default/classes/tasks/CollectionsTasks.cls:59]: good catch! it returned null. Added line 59, see last commit

File: force-app/main/default/classes/tasks/CollectionsTasks.cls [hunk 11]
```diff
@@ -7,11 +7,20 @@ public class CollectionsTasks {
...
-    for (Integer element : listWithZerosAndUnits) {
-      if (element == 0) {
-        countZeros++;
-      }
+    // no IFs solution
+    integer i = 0; //use for iteration in while loop
+    listWithZerosAndUnits.sort(); // sorted asc so that zeros are at the beginning of the list
+    system.debug('sorted: ' + listWithZerosAndUnits);
+    // iterate through the sorted list
+    while (i < listWithZerosAndUnits.size() && listWithZerosAndUnits[i] == 0) {
```
Comment (nchursin) [inline force-app/main/default/classes/tasks/CollectionsTasks.cls:22]: 👍 

not bad 😃 I didn't think of this solution, though it works too.

What I had in mind was using the fact it's a bit list, so it only has zeroes and ones. But your solution counts

File: force-app/main/default/classes/tasks/CollectionsTasks.cls [hunk 12]
```diff
@@ -46,7 +55,8 @@ public class CollectionsTasks {
       if (setFromList.contains(sum - element)) {
         result = true;
         break;
-      }
+      } else
+        result = false;
```
Comment (nchursin) [inline force-app/main/default/classes/tasks/CollectionsTasks.cls:67]: No, it's not that easy :)

Let's take a look at a more complex example:

```
Boolean result = CollectionsTasks.sumCanBeReachedOptimal(0, new List<Integer>{12, -3, 10, 1, 0});
System.debug('Should be false. Actual: ' + result);
```

 This also doesn't work

Comment (nchursin) [reply force-app/main/default/classes/tasks/CollectionsTasks.cls:67]: We will probably update the test with these examples... Cause other students have tried to hack the solution but faced similar issues

**Review by nchursin — CHANGES_REQUESTED** [review]

File: force-app/main/default/classes/tasks/CollectionsTasks.cls (hunk 11 above)
Comment (anna-blumenkrantz) [reply force-app/main/default/classes/tasks/CollectionsTasks.cls:22]: added also sum up bits solution

File: force-app/main/default/classes/tasks/CollectionsTasks.cls (hunk 12 above)
Comment (anna-blumenkrantz) [reply force-app/main/default/classes/tasks/CollectionsTasks.cls:67]: fixed by removing the element itself from checking .contains

Comment (nchursin) [reply force-app/main/default/classes/tasks/CollectionsTasks.cls:67]: 👍

**Review by nchursin — APPROVED** [review]

**Comment by nchursin** [comment]
DONE! 🎉

**Pull request closed without merging**